### DISCUSSION TITLE
SPI.transfer() hangs

### DISCUSSION THREAD

**Discussion by cpq** [earlephilhower/arduino-pico#2633] [Q&A] [answered]
I am about to use SPI module with my Arduino sketch, but what I noticed that SPI.transfer() hangs for some reason. Here is my sketch:

```c++
void setup() {
  Serial.begin(115200);
  while (!Serial) delay(10);

  SPI.begin();

  Serial.println("aa");   // I can see this message
  SPI.transfer(0);
  Serial.println("bb");   // I don't see this message
}
```

I can see the "aa" in the output, but not the "bb" - apparently  `SPI.transfer(0)` just hangs. **I do not have actual SPI  hardware connected**: testing with the bare, vanilla RP2040 Pico and RP2350 Pico2 boards. Both behave the same.

What am I doing wrong? Advice would be much appreciated. Thank you.

**Comment by earlephilhower** [comment, marked as answer]
In most modern Arduinos, you need to use SPI transactions.  Try adding `SPI.beginTransaction()` and `spi.endTransaction()` calls around the transfer(s).

**Reply by cpq** [reply]
@earlephilhower that was it, thank you so much!
